statsmodels/regression/mixed_glm.py

```diff
@@ -0,0 +1,1514 @@
+        omethod : string
+            The optimization method for scipy.optimize.minimize.
+
+        Returns
+        -------
+        float
+            The the marginal log-likelihood. We marginalize out the
+            random effects.
+        """
+
+        # The max of the function in the exponent of our integral
+        f, mp = self._get_map(fe_params, cov_re, scale, omethod)
+
+        like_grad_hess = self._joint_like_grad_hess(fe_params, cov_re, scale, ref = mp, fn=False, grad = False)
+        h = like_grad_hess[2]
+
+        d = len(mp)
+
+        # This is the log of the Laplace approximation.
+        # Algebraically we take the log of integral's approximation to avoid
+        # problems with numpy overflow with expressions like np.log(np.exp(.))
+        return - f + (d / 2.0) * np.log(2 * np.pi) - (1 / 2.0) * h
+
+    
+    def _joint_like_grad_hess(self, fe_params, cov_re, scale, ref, fn = True, grad = True, hess = True):
+        """
+        Function, gradient and hessian of the joint log likelihood.
+
+        The log-likelihood is for the data and random effects, viewed as a
+        function of the random effects.
+
+        Parameters
+        ----------
+        params : array-like, 1d
+            The fixed effects parameters.
+        cov_re : array-like, 2d
+            The covariance matrix of the random effects.
+        scale : float
+            The overdispersion scale parameter for the exponential family.
+        ref_re: array-like 2d
+
+        Returns
+        -------
+        tuple
+            the value of the negative joint log likelihood,
+            its gradient and the log-determinant of it's hessian
+            evaluated at the given state, respectively, in a tuple.
+        """
+
+        return (self._gen_joint_like_grad_hess(fe_params, cov_re, scale, fn, grad, hess))(ref)
+
+    
+    def _gen_joint_like_grad_hess(self, fe_params, cov_re, scale, fn = True, grad = True, hess = True):
+        """
+        Function, gradient and hessian of the joint log likelihood.
+
+        The log-likelihood is for the data and random effects, viewed as a
+        function of the random effects.
+
+        Parameters
+        ----------
+        params : array-like, 1d
+            The fixed effects parameters.
+        cov_re : array-like, 2d
+            The covariance matrix of the random effects.
+        scale : float
+            The overdispersion scale parameter for the exponential family.
+        fn: bool, optional
+            If True the returned function calculates the loglikelihood value.
+            Else the returned function always reports zero for the loglike.
+            If not specified defaults to True
+        grad: bool, optional
+            If True the returned function calculates the gradient value.
+            Else the returned function always reports a matrix of zeroes
+            for the gradient.
+            If not specified defaults to True
+        hess: bool, optional
```

Comment:
call it `hess_logdet` or something else, I got confused because I thought `hess` is hessian.